iarc_forebrain/scripts/forebrain.py

```diff
@@ -0,0 +1,175 @@
+import tf.transformations
+
+from math import pi
+
+
+class Strategy(object):
+    def __init__(self):
+        rospy.init_node('forebrain')
+
+        tfl = TransformListener()
+        self.drone = Drone(tfl=tfl)
+        self.world = WorldState(tfl=tfl)
+
+        rospy.sleep(1)
+
+    def test_hover(self):
+        self.world.waitForStart()
+
+        rospy.loginfo('Taking off')
+        self.drone.takeoff(1.5)
+
+        rospy.loginfo('Hovering')
+        self.drone.hover(5)
+
+        rospy.loginfo('Landing')
+        self.drone.land()
+
+        rospy.loginfo('Done!')
+
+    def test_square(self):
+        self.world.waitForStart()
+
+        rospy.loginfo('Taking off')
+        self.drone.takeoff(1.5)
+
+        rospy.loginfo('Hovering')
+        self.drone.hover(2)
+
+        for target in [(0, 1), (1, 1), (1, 0), (0, 0)]:
+            rospy.loginfo('Moving to {}'.format(target))
+            self.drone.move_to(target[0], target[1], 'odom')
+            rospy.loginfo('Hovering')
+            self.drone.hover(2)
+
+        rospy.loginfo('Landing')
+        self.drone.land()
+
+    def test_follow(self):
+        self.world.waitForStart()
+        r = rospy.Rate(20)
+
+        self.drone.takeoff(1.5)
+        while not rospy.is_shutdown():
+            target = self.choose_target(self.world.targets)
+            if target is not None:
+                self.drone.move_towards(0, 0, target.frame_id)
+            else:
+                self.drone.hover(0)
+            r.sleep()
+
+    def test_follow_redirect(self):
+        self.world.waitForStart()
+        r = rospy.Rate(20)
+
+        self.drone.takeoff(1.5)
+        while not rospy.is_shutdown():
+            target = self.choose_target(self.world.targets)
+            if target is not None:
+                if abs(angle_diff(self.world.target_facing_angle(target), self.world.CORRECT_DIRECTION)) > pi/2:
+                    success = self.drone.redirect_180(target)
+                    if success:
+                        rospy.loginfo('Redirected roomba: Success!')
+                    else:
+                        rospy.loginfo('Redirected roomba: Failure :(')
+                else:
+                    # Follow the roomba
+                    self.drone.move_towards(0, 0, target.frame_id, 1.5)
+            else:
+                self.drone.hover(0, 1.5)
+            r.sleep()
+
+    def choose_target(self, targets):
+        """
+        Selects the most important target
+        TODO: Use something more sophisticated
+        :param List[Roomba] targets:
+        :rtype: Roomba|None
+        """
+        if len(targets) > 0:
+            return targets[0]
+        return None
+
+    def run(self):
+        self.test_follow()
+
+
+def angle_diff(a, b):
+    """Returns x=a-b, wrapped to be -pi<x<=pi"""
+    diff = a - b
+    while diff <= -pi:
+        diff += 2 * pi
+    while diff > pi:
+        diff -= 2 * pi
+
+    return diff
+
+
+class WorldState(object):
```

Comment:
I suggest keeping a separate node explicitly targetted at keeping track of the world states.
I understand that this is part of a forebrain's job, so it may feel more natural to keep within a single node.

Reply:
It's not a separate node, that should be commented.